### ISSUE TITLE
API crashes after /medic/_bulk_docs gets called

### ISSUE THREAD

**Issue by henokgetachew** [medic/medic-webapp#3268]
Instance: cic-guatemalav2.dev.medicmobile.org
Web App version: 2.10.1
```
2017-03-21T23:32:35.687Z - info: 127.0.0.1 - - [21/Mar/2017:23:32:31 +0000] "POST /medic/_bulk_docs HTTP/1.0" - - "https://cic-guatemalav2.dev.medicmobile.org/medic/_design/medic/_rewrite/" "Mozilla/5.0 (Macintos
                                                                                                                                                                                                                    
2017-03-21T23:33:13.421Z - error: UNCAUGHT EXCEPTION!                                                                                                                                                               
                                                                                                                                                                                                                    
2017-03-21T23:33:13.424Z - error: Server error:                                                                                                                                                                     
 { [RangeError: Invalid string length]                                                                                                                                                                              
  domain:                                                                                                                                                                                                           
   { domain: null,                                                                                                                                                                                                  
     _events: { error: [Function] },                                                                                                                                                                                
     _maxListeners: undefined,                                                                                                                                                                                      
     members: [] },                                                                                                                                                                                                 
  domainThrown: true }                                                                                                                                                                                              
                                                                                                                                                                                                                    
2017-03-21T23:33:13.446Z - error: undefined                                                                                                                                                                         
                                                                                                                                                                                                                    
2017-03-21T23:33:16.906Z - info: DDoc extraction completed successfully                                                                                                                                             
                                                                                                                                                                                                                    
2017-03-21T23:33:17.164Z - info: Configuration loaded successfully                                                                                                                                                  
                                                                                                                                                                                                                    
2017-03-21T23:33:19.020Z - info: Translations merged successfully                                                                                                                                                   
                                                                                                                                                                                                                    
2017-03-21T23:33:19.161Z - info: Database migrations completed successfully                                                                                                                                         
                                                                                                                                                                                                                    
2017-03-21T23:33:19.206Z - info: Medic API listening on port 5988                                                                                                                                                   

```
Some more information: @guptasanchay was reporting that replication wasn't working for him as expected.  Docs that were deleted in admin are still showing up when logged in as the CHW.
Last replication date to the server has been showing as unknown.  Last replication from the server is updating properly (Logged in as CHW)

When logged in as a branch manager, replication to and from the server both look good

**Comment by garethbowen** [comment]
It's not necessarily after `_bulk_docs` - I've found all sorts of examples. I can't reproduce it however so I suspect there is a specific action causing the issue. @guptasanchay Is there anything you can do to reliably reproduce this issue? Which user are you logged in as?

I'll try and improve the logging so if it happens again we'll get more info.

**Comment by henokgetachew** [comment]
After improved logging:
```
2017-03-23T05:11:32.008Z - error: UNCAUGHT EXCEPTION!                                                                                                                                                                                         
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.008Z - error: ~~~~~~~~~~ SERVER ERROR                                                                                                                                                                                     
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.011Z - info: { [RangeError: Invalid string length]                                                                                                                                                                        
  domain:                                                                                                                                                                                                                                     
   { domain: null,                                                                                                                                                                                                                            
     _events: { error: [Function] },                                                                                                                                                                                                          
     _maxListeners: undefined,                                                                                                                                                                                                                
     members: [] },                                                                                                                                                                                                                           
  domainThrown: true }                                                                                                                                                                                                                        
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.012Z - info: ~~~~~~~~~~ STACK                                                                                                                                                                                             
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.013Z - info: RangeError: Invalid string length                                                                                                                                                                            
    at join (native)                                                                                                                                                                                                                          
    at Object.stringify (native)                                                                                                                                                                                                              
    at relax (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/nano/lib/nano.js:169:25)                                                        
    at Object.bulksDoc (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/nano/lib/nano.js:579:14)                                              
    at /srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/couchdb-audit/couchdb-audit/log.js:79:21                                               
    at /srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/async/dist/async.js:1012:9                                                             
    at /srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/async/dist/async.js:359:16                                                             
    at iteratorCallback (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/async/dist/async.js:935:13)                                          
    at /srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/async/dist/async.js:843:16                                                             
    at /srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/async/dist/async.js:1009:13                                                            
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.013Z - info: ~~~~~~~~~~ URL                                                                                                                                                                                               
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.013Z - info: /medic/_bulk_docs                                                                                                                                                                                            
                                                                                                                                                                                                                                              
2017-03-23T05:11:32.033Z - error: undefined                                                                                                                                                                                                   
                                                                                                                                                                                                                                              
2017-03-23T05:11:35.968Z - info: DDoc extraction completed successfully                                                                                                                                                                       
                                                                                                                                                                                                                                              
2017-03-23T05:11:36.373Z - info: Configuration loaded successfully                                                                                                                                                                            
                                                                                                                                                                                                                                              
2017-03-23T05:11:40.045Z - info: Translations merged successfully                                                                                                                                                                             
                                                                                                                                                                                                                                              
2017-03-23T05:11:40.317Z - info: Database migrations completed successfully                                                                                                                                                                   
                                                                                                                                                                                                                                              
2017-03-23T05:11:40.398Z - info: Medic API listening on port 5988                                                                                                                                                                             

```

**Comment by henokgetachew** [comment]
It seems like it's related to this: https://github.com/medic/couchdb-audit/blob/master/couchdb-audit/log.js#L79

**Comment by garethbowen** [comment]
Ok, it looks like it's a really bad explanation from v8 when you're trying to `stringify` too much. Either you have heaps of unreplicated data (lots of docs, or one massive doc) that's all trying to go out in one replication cycle, or you've got one massive audit record that can't be updated any more. I'll do some more investigation to try and work out which option it is...

Also, you're running an ancient version of node which probably isn't helping.

**Comment by garethbowen** [comment]
I found a massive doc (~2GB) in medic-audit db which is the `medic` ddoc. The medic ddoc should never be replicated back to the server so this is definitely a bug. As far as I can tell this restricted user downloaded the ddoc correctly in v2.9 but then after upgrade to v2.10 kept trying to sync it back.

This will only affect clients who have loaded the app in v2.9 and then upgraded to 2.10. @guptasanchay managed to work around the issue by clearing his local pouchdb and replicating again. I've deleted the ddoc audit doc to clear up some space. 

I'll patch the code so ddocs are never replicated back to the server and add to the 2.10.2 release so nobody else is affected by this issue.
